### ISSUE TITLE
NeurIPS 2023 | Rewiring Neurons in Non-Stationary Environments

### ISSUE THREAD

**Issue by chufanchen** [chufanchen/read-paper-and-code#70]
https://github.com/feifeiobama/RewireNeuron

https://proceedings.neurips.cc/paper_files/paper/2023/file/599221d7ebf6b3403190f38a3f282a1c-Paper-Conference.pdf

Related work:

SoftSort: A Continuous Relaxation for the argsort Operator

**Comment by ChufanSuki** [comment]
### Problem

Continual RL, non-stationary environments, $T$ tasks.

Goal: global policy $\pi(a\vert s, t)$, task identifier $t \in [1,\dots, T]$.

Policy network: MLP, affine layers $W=\{W_l\}_{l=1}^{L}$, element-wise activation functions $\sigma$. 

$$
Y=W_L \circ \sigma \circ W_{L-1} \circ \dots \circ \sigma \circ W_1 X
$$

where $X$ and $Y$ denote the input and the predicted policy.


### Rewiring via permutation

permutation matrices $P=\{P_l\}_{l=1}^L$

$$
Y=W_L \circ \sigma \circ P_{L-1} W_{L-1} \circ \ldots \circ \sigma \circ P_1 W_1 X
$$

**Comment by ChufanSuki** [comment]
### Rewiring for exploration

We propose a multi-mode rewiring strategy that maintains a set of different wirings to induce diverse policies. t each step, the agent randomly samples a mode to select from the following candidate wirings:

$$
P_l \in \{P_{l, 1}, P_{l, 2}, \ldots, P_{l, K} \}
$$

where $K$ is the total number of modes, and the mode is shared across all layers. Also, to ensure an equivalent learning rate, we multiply the learning rate of each wiring by $K$.

We further distill knowledge across different wirings for knowledge sharing, which resembles the idea of mutual learning but in a parameter-efficient manner. Kullback-Leibler (KL) divergence is used as the loss:

$$
L_{\mathrm{KL}}(W, P)=\mathbb{E}_{k^{\prime} \neq k}[D_{\mathrm{KL}}\left(\pi_{k^{\prime}}(\cdot \mid \mathbf{s}) \Vert \pi_k(\cdot \mid \mathbf{s})\right)],
$$

where $\pi_k$ is the policy induced by the $k$-th wiring, and we only backpropagate through one wiring to save computational cost.

### Rewiring for stability-plasticity

#### Caching each wiring

cache each learned wiring and regularize the weights

$$
L_{\mathrm{reg}}\left(\boldsymbol{W}^t\right)=\sum_{l=1}^L\left\Vert\boldsymbol{W}_l^t-\boldsymbol{W}_l^{t-1}\right\Vert^2
$$

###$ Aligning wiring with weights

misalignment between latest weights $W_l^t$ and cached wirings $P_l^{t-1}$

$$
L_{\mathrm{SP}}\left(\boldsymbol{W}^t, \boldsymbol{P}^{\prime}, \boldsymbol{P}^{\prime \prime}\right)=\sum_{l=1}^L\left\Vert\boldsymbol{W}_l^t-\boldsymbol{P}_l^{\prime \prime} \boldsymbol{W}_l^{t-1} \boldsymbol{P}_{l-1}^{\prime \top}\right\Vert^2
$$

$$
\boldsymbol{P}_l^{t-1}=\boldsymbol{P}_l^{\prime} \boldsymbol{P}_l^{t-1} \boldsymbol{P}_l^{\prime \prime \top}
$$
